lib/myapp18_web/user_auth.ex

```diff
@@ -98,13 +110,21 @@ defmodule Myapp18Web.UserAuth do
   end
 
   @doc """
-  Authenticates the user by looking into the session
-  and remember me token.
+  Authenticates the user by looking into the session and remember me token.
+
+  Will reissue the session token if it is older than the configured age.
   """
   def fetch_current_scope_for_user(conn, _opts) do
-    {user_token, conn} = ensure_user_token(conn)
-    user = user_token && Accounts.get_user_by_session_token(user_token)
-    assign(conn, :current_scope, Scope.for_user(user))
+    with(
+      {token, conn} <- ensure_user_token(conn),
+      {user, token_created} <- Accounts.get_user_by_session_token(token)
+    ) do
+      conn
+      |> assign(:current_scope, Scope.for_user(user))
+      |> maybe_reissue_user_session_token(user, token, token_created)
+    else
+      nil -> assign(conn, :current_scope, Scope.for_user(nil))
+    end
```

Comment:
Given the first one always succeeds, there is really only one branch here, so we can use `case`:

```suggestion
    {token, conn} = ensure_user_token(conn)

    case Accounts.get_user_by_session_token(token) do
      {user, token_created} ->
        conn
        |> assign(:current_scope, Scope.for_user(user))
        |> maybe_reissue_user_session_token(user, token, token_created)

      nil ->
        assign(conn, :current_scope, Scope.for_user(nil))
    end
```

Reply:
The first case in the `with` doesn't always succeed if there is no token in the session or cookie. The original `ensure_user_token(conn)` would return `{nil, conn}` in that case, but I modified it to return `nil` instead to go along with the other condition in the `with` and keep LOC to a minimum.

Without the change to `ensure_user_token`, the `case` would need to be wrapped in a `if token do` or else make `Accounts.get_user_by_session_token` handle nil tokens as input, which no other function in `Accounts` does, so I didn't think this was an option.

Other options that pass all tests along with `ensure_user_token` returning `{nil, conn}` that I can think of:

```elixir
  def fetch_current_scope_for_user(conn, _opts) do
    {token, conn} = ensure_user_token(conn)

    if token do
      case Accounts.get_user_by_session_token(token) do
        {user, token_inserted_at} ->
          conn
          |> assign(:current_scope, Scope.for_user(user))
          |> maybe_reissue_user_session_token(user, token, token_inserted_at)

        nil ->
          assign(conn, :current_scope, Scope.for_user(nil))
      end
    else
      assign(conn, :current_scope, Scope.for_user(nil))
    end
  end
```

or

```elixir
  def fetch_current_scope_for_user(conn, _opts) do
    {token, conn} = ensure_user_token(conn)

    {user, conn} =
      if token do
        if {user, token_inserted_at} = Accounts.get_user_by_session_token(token) do
          {user, maybe_reissue_user_session_token(conn, user, token, token_inserted_at)}
        else
          {nil, conn}
        end
      else
        {nil, conn}
      end

    assign(conn, :current_scope, Scope.for_user(user))
  end
```

Unless you can think of something else.

If it was my code, I would add `def get_user_by_session_token(nil), do: nil` to `Accounts` and call it a day. But I honestly don't know if you would like that or not, and there's no other example of it there already.

Reply:
Ignore me. Your code is better then :)